service/worker/migration/activities.go

```diff
@@ -252,24 +254,37 @@ func (a *activities) checkHandoverOnce(ctx context.Context, waitRequest waitHand
 	return readyShardCount == len(resp.Shards), nil
 }
 
-func (a *activities) generateWorkflowReplicationTask(ctx context.Context, wKey definition.WorkflowKey) error {
+func (a *activities) generateWorkflowReplicationTask(ctx context.Context, rateLimiter quotas.RateLimiter, wKey definition.WorkflowKey) error {
+	if err := rateLimiter.WaitN(ctx, 1); err != nil {
+		return err
+	}
+
 	// will generate replication task
 	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
 	defer cancel()
 
-	_, err := a.historyClient.GenerateLastHistoryReplicationTasks(ctx, &historyservice.GenerateLastHistoryReplicationTasksRequest{
+	resp, err := a.historyClient.GenerateLastHistoryReplicationTasks(ctx, &historyservice.GenerateLastHistoryReplicationTasksRequest{
 		NamespaceId: wKey.NamespaceID,
 		Execution: &commonpb.WorkflowExecution{
 			WorkflowId: wKey.WorkflowID,
 			RunId:      wKey.RunID,
 		},
 	})
 
-	if _, isNotFound := err.(*serviceerror.NotFound); isNotFound {
-		// ignore NotFound error
+	switch err.(type) {
+	case nil:
+		stateTransitionCount := resp.StateTransitionCount
+		for stateTransitionCount > 0 {
+			token := util.Min(int(stateTransitionCount), rateLimiter.Burst())
+			stateTransitionCount -= int64(token)
+			_ = rateLimiter.ReserveN(time.Now(), token)
```

Comment:
are you still intend to wait on the token? I think `ReserveN` only returns reservation. You still need to call `time.Sleep(r.Delay())` to wait according to the doc.

Reply:
since we need to call the API to know the # of LWTs (tokens to consume)
here the logic should just reserve those tokens, next call (WaitN function) will be blocked

Reply:
I c. Can you add some comment to help understanding? Are you suggesting `WaitN(, 1)` at line 258 will wait on all reserved tokens? WaitN comment says "WaitN blocks until lim permits n events to happen." so I assume it will just consume 1 token?

Reply:
>Are you suggesting WaitN(, 1) at line 258 will wait on all reserved tokens?

plz take a look at the rate limiter doc: https://pkg.go.dev/golang.org/x/time/rate#Limiter.ReserveN